proposals/conversion-webhook.md

```diff
@@ -0,0 +1,73 @@
+# Conversion Webhook
+
+- **Authors**: [Mariam Fahmy](https://github.com/MariamFahmy98)
+- **Created**: December 12th, 2023
+- **Abstract**: Implementing a conversion webhook to convert from v1 to v2 and vice versa.
+
+## Overview
+- [Introduction](#introduction)
+- [Motivation](#motivation)
+- [Proposal](#proposal)
+- [Implementation](#implementation)
+- [Next Steps](#next-steps)
+
+## Introduction
+Conversion webhooks are a feature introduced in Kubernetes to facilitate the transformation and validation of custom resources. They enable automatic conversion of incoming requests between different versions of the CRD schema. This ensures backward compatibility as the CRD evolves over time, allowing existing resources to be seamlessly upgraded without disruption.
+
+Some common use cases where conversion is required:
+1. custom resource is requested in a different version than stored version.
+2. Watch is created in one version but the changed object is stored in another version.
+3. custom resource PUT request is in a different version than storage version.
+
+## Motivation
+Kyverno CRDs `v1` contain many deprecated fields that require removal. Additionally, there are fields at the `spec` level that are only specific to certain rules and not others. Therefore, we would like to introduce a more stable version, `v2`. Since `v1` and `v2` will have different schema, a conversion webhook is required to convert between them.
+
+## Proposal
+In Kubernetes, it is essential that all versions are compatible with each other. This means that if we go from `v1` to `v2` and then back to `v1`, we must not lose any information. Therefore, any changes we make to our API must be compatible with `v1`, and we must ensure that anything we add in `v2` is also supported in `v1`. This requires us to add new fields to `v1` to support any new additions.
```

Comment:
> This requires us to add new fields to `v1` to support any new additions.

I don't quite understand this, isn't it possible for us to use an annotation (as stated in solution 2.) also on the V1 object? So if something is only available in v2, it gets written to an annotation in v1 and then we don't lose data. Maybe I am missing something why this wouldn't work.

Reply:
We can do that as well.